src/ping.c

```diff
@@ -431,151 +486,188 @@ static int ping_init(void) /* {{{ */
+    } else if (strcasecmp(child->key, "Size") == 0) {
+      int packet_size;
+      if (cf_util_get_int(child, &packet_size) != 0)
+        return 1;
+      size_t size = (size_t)(packet_size);
+
+      /* Max IP packet size - (IPv6 + ICMP) = 65535 - (40 + 8) = 65487 */
+      if (size <= 65487) {
+        sfree(ping_data);
+        ping_data = malloc(size + 1);
+        if (ping_data == NULL) {
+          ERROR("ping plugin: malloc failed.");
+          return 1;
+        }
+
+        /* Note: By default oping is using constant string
+         * "liboping -- ICMP ping library <http://octo.it/liboping/>"
+         * which is exactly 56 bytes.
+         *
+         * Optimally we would follow the ping(1) behaviour, but we
+         * cannot use byte 00 or start data payload at exactly same
+         * location, due to oping library limitations. */
+        for (size_t i = 0; i < size; i++) /* {{{ */
+        {
+          /* This restricts data pattern to be only composed of easily
+           * printable characters, and not NUL character. */
+          ping_data[i] = ('0' + i % 64);
+        } /* }}} for (i = 0; i < size; i++) */
+        ping_data[size] = 0;
+      } else
+        WARNING("ping plugin: Ignoring invalid Size %" PRIsz ".", size);
+    } else if (strcasecmp(child->key, "Timeout") == 0) {
+      double tmp;
+      if (cf_util_get_double(child, &tmp) != 0)
+        return 1;
+      if (tmp > 0.0)
+        ping_timeout = tmp;
+      else
+        WARNING("ping plugin: Ignoring invalid timeout %g", tmp);
+    } else if (strcasecmp(child->key, "MaxMissed") == 0) {
+      if (cf_util_get_int(child, &ping_max_missed) != 0)
+        return 1;
 
-    tmp = atof(value);
-    if (tmp > 0.0)
-      ping_timeout = tmp;
-    else
-      WARNING("ping plugin: Ignoring invalid timeout %g (%s)", tmp, value);
-  } else if (strcasecmp(key, "MaxMissed") == 0) {
-    ping_max_missed = atoi(value);
-    if (ping_max_missed < 0)
-      INFO("ping plugin: MaxMissed < 0, disabled re-resolving of hosts");
-  } else {
-    return -1;
+      if (ping_max_missed < 0)
+        INFO("ping plugin: MaxMissed < 0, disabled re-resolving of hosts");
+    } else {
+      WARNING("ping plugin: The config option "
+              "\"%s\" is not allowed.", child->key);
+      return -1;
+    }
   }
-
   return 0;
 } /* }}} int ping_config */
 
-static void submit(const char *host, const char *type, /* {{{ */
-                   gauge_t value) {
+static void submit(const char *source, const char *host, const char *type, /* {{{ */
+    gauge_t value) {
   value_list_t vl = VALUE_LIST_INIT;
-
-  vl.values = &(value_t){.gauge = value};
-  vl.values_len = 1;
-  sstrncpy(vl.plugin, "ping", sizeof(vl.plugin));
-  sstrncpy(vl.type_instance, host, sizeof(vl.type_instance));
-  sstrncpy(vl.type, type, sizeof(vl.type));
-
-  plugin_dispatch_values(&vl);
+  char *src_host = (char *)malloc(strlen(source) + strlen(host) + 2);
```

Comment:
`src_host` is only written to, but never read. Are you missing a

```c
sstrncpy(vl.host, src_host, sizeof(vl.host));
```

in this function?